File: 2026/rtti/cpp/test/rtti_test.cpp
#include "rtti/rtti.h"

#include <cassert>
#include <iostream>
#include <memory>
#include <vector>

using namespace rtti;

// ---- Test model classes ----

struct Entity : Object {
    RTTI_ENABLE(Entity)
};

struct Player : Entity {
    RTTI_ENABLE(Player)
};

struct Enemy : Entity {
    RTTI_ENABLE(Enemy)
};

struct Orc : Enemy {
    RTTI_ENABLE(Orc)
};

struct Dragon : Enemy {
    RTTI_ENABLE(Dragon)
};

// ---- Registration ----

static void register_types() {
    auto& reg = type_registry();

    auto* entity  = reg.define("Entity");
    auto* player  = reg.define("Player", entity);
    auto* enemy   = reg.define("Enemy", entity);
    auto* orc     = reg.define("Orc", enemy);
    auto* dragon  = reg.define("Dragon", enemy);

    reg.seal();
}

// ---- Tests ----

static void test_interval_assignment() {
    auto& reg = type_registry();

    const TypeInfo* entity = reg.lookup("Entity");
    const TypeInfo* player = reg.lookup("Player");
    const TypeInfo* enemy  = reg.lookup("Enemy");
    const TypeInfo* orc    = reg.lookup("Orc");
    const TypeInfo* dragon = reg.lookup("Dragon");

    assert(entity->id == 1 && entity->high == 5);
    assert(enemy->id == 2 && enemy->high == 4);
    assert(dragon->id == 3 && dragon->high == 3);
    assert(orc->id == 4 && orc->high == 4);
    assert(player->id == 5 && player->high == 5);

    std::cout << "  [PASS] test_interval_assignment\n";
}

static void test_is_assignable_from() {
    const TypeInfo* entity = type_registry().lookup("Entity");
    const TypeInfo* player = type_registry().lookup("Player");
    const TypeInfo* enemy  = type_registry().lookup("Enemy");
    const TypeInfo* orc    = type_registry().lookup("Orc");
    const TypeInfo* dragon = type_registry().lookup("Dragon");

    assert(entity->is_assignable_from(*player));
    assert(entity->is_assignable_from(*enemy));
    assert(entity->is_assignable_from(*orc));
    assert(entity->is_assignable_from(*dragon));

    assert(!player->is_assignable_from(*entity));
    assert(!player->is_assignable_from(*enemy));
    assert(!player->is_assignable_from(*orc));

    assert(enemy->is_assignable_from(*orc));
    assert(enemy->is_assignable_from(*dragon));
    assert(!enemy->is_assignable_from(*entity));
    assert(!enemy->is_assignable_from(*player));

    std::cout << "  [PASS] test_is_assignable_from\n";
}

static void test_object_type_check() {
    Orc orc;
    const TypeInfo* entity = type_registry().lookup("Entity");
    const TypeInfo* player = type_registry().lookup("Player");
    const TypeInfo* enemy  = type_registry().lookup("Enemy");
    const TypeInfo* orc_t  = type_registry().lookup("Orc");
    const TypeInfo* dragon = type_registry().lookup("Dragon");

    assert(orc.is_instance(*entity));
    assert(!orc.is_instance(*player));
    assert(orc.is_instance(*enemy));
    assert(orc.is_instance(*orc_t));
    assert(!orc.is_instance(*dragon));

    assert(orc.type_of().name == "Orc");

    std::cout << "  [PASS] test_object_type_check\n";
}

static void test_cast_success() {
    Orc orc;
    Entity* e = &orc;

    Orc* o = rtti::cast<Orc>(e);
    assert(o != nullptr);

    std::cout << "  [PASS] test_cast_success\n";
}

static void test_cast_fail() {
    Dragon dragon;
    Entity* e = &dragon;

    bool caught = false;
    try {
        rtti::cast<Orc>(e);
    } catch (const std::bad_cast&) {
        caught = true;
    }
    assert(caught);

    std::cout << "  [PASS] test_cast_fail\n";
}

static void test_is_template() {
    Orc orc;
    assert(orc.is<Orc>());
    assert(orc.is<Enemy>());
    assert(orc.is<Entity>());
    assert(!orc.is<Player>());
    assert(!orc.is<Dragon>());

    std::cout << "  [PASS] test_is_template\n";
}

static void test_huge_hierarchy() {
    TypeRegistry reg;

    TypeInfo* root = reg.define("Root");
    for (int i = 0; i < 1000; i++) {
        reg.define("Child" + std::to_string(i), root);
    }
    reg.seal();

    const TypeInfo* child0 = reg.lookup("Child0");
    assert(child0->id == 2);
    assert(child0->high == 2);
    assert(root->is_assignable_from(*child0));
    assert(!child0->is_assignable_from(*root));

    std::cout << "  [PASS] test_huge_hierarchy\n";
}

static void test_rtti_utility() {
    Player player;
    const TypeInfo* entity = type_registry().lookup("Entity");
    const TypeInfo* enemy  = type_registry().lookup("Enemy");

    assert(rtti::is_instance(&player, *entity));
    assert(!rtti::is_instance(&player, *enemy));
    assert(rtti::type_of(&player)->name == "Player");

    std::cout << "  [PASS] test_rtti_utility\n";
}

int main() {
    register_types();

    std::cout << "Running C++ RTTI tests...\n";
    test_interval_assignment();
    test_is_assignable_from();
    test_object_type_check();
    test_cast_success();
    test_cast_fail();
    test_is_template();
    test_huge_hierarchy();
    test_rtti_utility();
    std::cout << "All tests passed.\n";

    return 0;
}

File: 2026/rtti/java/src/test/java/rtti/RttiTest.java
package rtti;

import org.junit.jupiter.api.BeforeAll;
import org.junit.jupiter.api.Test;

import static org.junit.jupiter.api.Assertions.*;

class RttiTest {

    static TypeInfo entityType;
    static TypeInfo playerType;
    static TypeInfo enemyType;
    static TypeInfo orcType;
    static TypeInfo dragonType;

    @BeforeAll
    static void setup() {
        TypeRegistry reg = TypeRegistry.get();

        entityType = reg.define("Entity", null, Entity.class);
        playerType = reg.define("Player", entityType, Player.class);
        enemyType = reg.define("Enemy", entityType, Enemy.class);
        orcType = reg.define("Orc", enemyType, Orc.class);
        dragonType = reg.define("Dragon", enemyType, Dragon.class);

        reg.seal();
    }

    @Test
    void testIntervalAssignment() {
        assertEquals(1, entityType.id());
        assertEquals(5, entityType.high());

        assertEquals(2, playerType.id());
        assertEquals(2, playerType.high());

        assertEquals(3, enemyType.id());
        assertEquals(5, enemyType.high());

        assertEquals(4, orcType.id());
        assertEquals(4, orcType.high());

        assertEquals(5, dragonType.id());
        assertEquals(5, dragonType.high());
    }

    @Test
    void testIsAssignableFrom() {
        assertTrue(entityType.isAssignableFrom(playerType));
        assertTrue(entityType.isAssignableFrom(enemyType));
        assertTrue(entityType.isAssignableFrom(orcType));
        assertTrue(entityType.isAssignableFrom(dragonType));

        assertFalse(playerType.isAssignableFrom(entityType));
        assertFalse(playerType.isAssignableFrom(enemyType));
        assertFalse(playerType.isAssignableFrom(orcType));

        assertTrue(enemyType.isAssignableFrom(orcType));
        assertTrue(enemyType.isAssignableFrom(dragonType));
        assertFalse(enemyType.isAssignableFrom(entityType));
        assertFalse(enemyType.isAssignableFrom(playerType));
    }

    @Test
    void testObjectTypeCheck() {
        RttiObject orc = new Orc(orcType);

        assertTrue(orc.isInstance(entityType));
        assertFalse(orc.isInstance(playerType));
        assertTrue(orc.isInstance(enemyType));
        assertTrue(orc.isInstance(orcType));
        assertFalse(orc.isInstance(dragonType));

        assertEquals("Orc", orc.typeOf().name());
    }

    @Test
    void testCastSuccess() {
        Enemy enemy = new Orc(orcType);

        Orc orc = enemy.cast(Orc.class);
        assertNotNull(orc);
    }

    @Test
    void testCastFail() {
        Enemy enemy = new Dragon(dragonType);

        assertThrows(ClassCastException.class, () -> enemy.cast(Orc.class));
    }

    @Test
    void testLookup() {
        assertEquals("Entity", TypeRegistry.get().lookup(1).name());
        assertEquals("Orc", TypeRegistry.get().lookup(4).name());
        assertEquals("Dragon", TypeRegistry.get().lookup("Dragon").name());
        assertEquals(Entity.class, TypeRegistry.get().lookup(Entity.class).javaClass());
    }

    @Test
    void testRttiUtil() {
        Player player = new Player(playerType);
        assertTrue(Rtti.isInstance(player, entityType));
        assertFalse(Rtti.isInstance(player, enemyType));
        assertEquals("Player", Rtti.typeOf(player).name());
    }

    @Test
    void testHugeHierarchy() {
        TypeRegistry reg = new TypeRegistry();

        TypeInfo root = reg.define("Root", null);
        for (int i = 0; i < 1000; i++) {
            reg.define("Child" + i, root);
        }
        reg.seal();

        TypeInfo child0 = reg.lookup("Child0");
        assertEquals(2, child0.id());
        assertEquals(2, child0.high());
        assertTrue(root.isAssignableFrom(child0));
        assertFalse(child0.isAssignableFrom(root));
    }

    // ---- Test model classes (no static TYPE fields) ----

    static class Entity extends RttiObject {
        Entity() { super(0); }
    }

    static class Player extends Entity {
        Player(TypeInfo t) { super(t); }
    }

    static class Enemy extends Entity {
        Enemy(TypeInfo t) { super(t); }
    }

    static class Orc extends Enemy {
        Orc(TypeInfo t) { super(t); }
    }

    static class Dragon extends Enemy {
        Dragon(TypeInfo t) { super(t); }
    }
}
